compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -1033,6 +1033,16 @@ impl<'a> Parser<'a> {
         if self.eat_keyword(kw::For) {
             self.expect_lt()?;
             let params = self.parse_generic_params()?;
+            let params = params
+                .into_iter()
+                .map(|param| match param.kind {
+                    GenericParamKind::Type { ref default } if let Some(ty) = default => {
+                        self.sess.emit_err(UnexpectedDefaultParameterInBinder { span: ty.span });
+                        GenericParam { kind: GenericParamKind::Type { default: None }, ..param }
+                    }
+                    _ => param,
+                })
```

Comment:
Rejecting defaults on higher-ranked generic parameters in the *parser* is not an option for backward compatibility reasons since `#[cfg(FALSE)] … for<T = ()> …` successfully compiles on stable and it it needs to stay that way.

I think what errs meant is rejecting them during AST Validation (→ `astvalidation::AstValidator`).